src/twisted/application/test/test_internet.py

```diff
@@ -860,3 +860,103 @@ def test_retryCancelled(self):
         d = service.stopService()
         self.assertEqual(clock.getDelayedCalls(), [])
         self.successResultOf(d)
+
+
+    def test_stopServiceBeforeStartService(self):
+        """
+        Calling L{ClientService.stopService} before
+        L{ClientService.startService} returns a L{Deferred} that has
+        already fired with L{None}.
+        """
+        clock = Clock()
+        _, service = self.makeReconnector(fireImmediately=False,
+                                          startService=False,
+                                          clock=clock)
+        d = service.stopService()
+        self.assertIsNone(self.successResultOf(d))
+
+
+    def test_whenConnectedErrbacksOnStopService(self):
+        """
+        L{ClientService.whenConnected} returns a L{Deferred} that
+        errbacks with L{CancelledError} if the connection fails before
```

Comment:
This could do with a better explanation. It is specifically when it is called between connection attempts or during the delay before reconnecting that this is testing.